### PR TITLE
Migrate to pyQuil v3

### PR DESCRIPTION
**Pull request by MarquessV**
Here are the changes required to get this plugin on to pyQuil v3! I'm including some of my own targeted questions as comments below.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -28,11 +28,11 @@ PennyLane Forest Plugin
 .. header-start-inclusion-marker-do-not-remove
 
 The PennyLane Forest plugin allows different Rigetti devices to work with
-PennyLane --- the wavefunction simulator, and the Quantum Virtual Machine (QVM).
+PennyLane --- the wavefunction simulator, and the Quantum Virtual Machine (QVM), and Quantum Processing Units (QPUs).
```
Comment (AlbertMitjans) [inline README.rst:31]:
```suggestion
PennyLane --- the wavefunction simulator, the Quantum Virtual Machine (QVM), and Quantum Processing Units (QPUs).
```

File: README.rst [hunk 2]
```diff
@@ -47,8 +47,8 @@ Features
 ========
 
 * Provides three devices to be used with PennyLane: ``forest.numpy_wavefunction``,
```
Comment (AlbertMitjans) [inline README.rst:49]:
```suggestion
* Provides four devices to be used with PennyLane: ``forest.numpy_wavefunction``,
```

File: README.rst [hunk 3]
```diff
@@ -47,8 +47,8 @@ Features
...
-  ``forest.wavefunction``, and ``forest.qvm``. These provide access to the pyQVM
-  Numpy wavefunction simulator, Forest wavefunction simulator, and quantum
+  ``forest.wavefunction``, ``forest.qvm``, and ``forest.qpu``. These provide access to the pyQVM
+  Numpy wavefunction simulator, pyQuil wavefunction simulator, and quantum
   virtual machine (QVM) respectively.
```
Comment (AlbertMitjans) [inline README.rst:52]:
```suggestion
  Numpy wavefunction simulator, pyQuil wavefunction simulator, quantum
  virtual machine (QVM), and quantum processing units (QPUs) respectively.
```

File: README.rst [hunk 4]
```diff
@@ -59,7 +59,7 @@ Features
   ``ISWAP``, ``PSWAP``, and ``CPHASE``. Every custom operation supports analytic
   differentiation.
```
Comment (AlbertMitjans) [inline README.rst:60]: These gates are now defined in PennyLane:
- `ISWAP`: 
https://github.com/PennyLaneAI/pennylane/blob/e79f264adf743dbd35f2811de367edb9789718c3/pennylane/ops/qubit/non_parametric_ops.py#L1426

- `PSWAP`: https://github.com/PennyLaneAI/pennylane/blob/e79f264adf743dbd35f2811de367edb9789718c3/pennylane/ops/qubit/parametric_ops.py#L3189

- `CPHASE`: 
https://github.com/PennyLaneAI/pennylane/blob/e79f264adf743dbd35f2811de367edb9789718c3/pennylane/ops/qubit/parametric_ops.py#L546

File: README.rst [hunk 5]
```diff
@@ -86,14 +86,12 @@ Dependencies
 
 PennyLane-Forest requires the following libraries be installed:
 
-* `Python <http://python.org/>`__ >=3.6
+* `Python <http://python.org/>`__ >=3.7
 
 as well as the following Python packages:
 
 * `PennyLane <http://pennylane.readthedocs.io/>`__
```
Comment (AlbertMitjans) [inline README.rst:93]: We may need to add a lower bound on the PennyLane version. Will look more into it.

**Review by AlbertMitjans — COMMENTED** [review]
Left some comments on the `README.rst` file

File: doc/devices/qpu.rst [hunk 6]
```diff
@@ -1,5 +1,3 @@
-:orphan:
-
 The QPU device
```
Comment (AlbertMitjans) [inline doc/devices/qpu.rst:1]: Delete duplicated `very similar` text in the first paragraph of this section

File: doc/devices/qpu.rst [hunk 7]
```diff
@@ -64,33 +62,18 @@ These operations include:
     pennylane_forest.ops.ISWAP
     pennylane_forest.ops.PSWAP
```
Comment (AlbertMitjans) [inline doc/devices/qpu.rst:63]: The three gates mentioned here are now supported in PennyLane. Maybe remove the last part of this section?

File: doc/devices/qpu.rst [hunk 8]
```diff
@@ -64,33 +62,18 @@ These operations include:
...
-
-``qvm_url`` (*str*)
-    the QVM server URL. Can also be set by the environment
-    variable ``QVM_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:5000"``.
-
-``compiler_url`` (*str*)
-    the compiler server URL. Can also be set by the environment
-    variable ``COMPILER_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:6000"``.
```
Comment (AlbertMitjans) [inline doc/devices/qpu.rst:86]: Why did you delete these keyword arguments? I believe you talk about them in the `quilc server configuration` right below.

File: doc/devices/qvm.rst [hunk 9]
```diff
@@ -190,36 +190,4 @@ After measuring the qubit state, we can determine the probability :math:`P_0` of
```
Comment (AlbertMitjans) [inline doc/devices/qvm.rst:190]: Consider removing the `CPHASE`, `ISWAP` and `PSWAP` gates from the `Supported operations` section of this file, given that they are already supported in PennyLane.

File: doc/devices/qvm.rst [hunk 10]
```diff
@@ -190,36 +190,4 @@ After measuring the qubit state, we can determine the probability :math:`P_0` of
...
-
-``qvm_url`` (*str*)
-    the QVM server URL. Can also be set by the environment
-    variable ``QVM_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:5000"``.
-
-``compiler_url`` (*str*)
-    the compiler server URL. Can also be set by the environment
-    variable ``COMPILER_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:6000"``.
```
Comment (AlbertMitjans) [inline doc/devices/qvm.rst:215]: Same as before, you delete these keyword arguments but then mention them in the next section.

File: doc/devices/wavefunction.rst [hunk 11]
```diff
@@ -50,37 +50,6 @@ You can then execute the circuit like any other function to get the quantum mech
...
-
-``qvm_url`` (*str*)
-    the QVM server URL. Can also be set by the environment
-    variable ``QVM_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:5000"``.
-
-``compiler_url`` (*str*)
-    the compiler server URL. Can also be set by the environment
-    variable ``COMPILER_URL``, or in the ``~/.forest_config`` configuration file.
-    Default value is ``"http://127.0.0.1:6000"``.
```
Comment (AlbertMitjans) [inline doc/devices/wavefunction.rst:72]: Same as before.

File: doc/devices/wavefunction.rst [hunk 12]
```diff
@@ -95,4 +64,6 @@ These operations include:
 .. autosummary::
     pennylane_forest.ops.CPHASE
     pennylane_forest.ops.ISWAP
-    pennylane_forest.ops.PSWAP
\ No newline at end of file
+    pennylane_forest.ops.PSWAP
```
Comment (AlbertMitjans) [inline doc/devices/wavefunction.rst]: Same: these operations are now supported in PennyLane

Comment (AlbertMitjans) [reply doc/devices/wavefunction.rst]: Consider deleting these ops also in the `numpy-wavefunction` section

File: doc/index.rst [hunk 13]
```diff
@@ -45,7 +50,7 @@ Check out these demos to see the PennyLane-Forest plugin in action:
     <div class="row">
 
 .. title-card::
```
Comment (AlbertMitjans) [inline doc/index.rst:52]: I think this `title-card` and the next one are not rendered properly (see [link](https://xanaduai-pennylane--107.com.readthedocs.build/projects/forest/en/107/index.html)).

**Review by AlbertMitjans — COMMENTED** [review]
Added more comments on the `doc` files

File: pennylane_forest/qpu.py [hunk 14]
```diff
@@ -66,28 +61,19 @@ class QPUDevice(QVMDevice):
...
-            the environment variable ``FOREST_SERVER_URL``, or in the ``~/.qcs_config``
-            configuration file. Default value is ``"https://forest-server.qcs.rigetti.com"``.
-        qvm_url (str): the QVM server URL. Can also be set by the environment
-            variable ``QVM_URL``, or in the ``~/.forest_config`` configuration file.
-            Default value is ``"http://127.0.0.1:5000"``.
-        compiler_url (str): the compiler server URL. Can also be set by the environment
-            variable ``COMPILER_URL``, or in the ``~/.forest_config`` configuration file.
-            Default value is ``"http://127.0.0.1:6000"``.
-        timeout (int): number of seconds to wait for a response from the client.
         parametric_compilation (bool): a boolean value of whether or not to use parametric
```
Comment (AlbertMitjans) [inline pennylane_forest/qpu.py:70]: Should we add `compiler_timeout` and `execution_timeout`?

File: pennylane_forest/qvm.py [hunk 15]
```diff
@@ -228,14 +236,19 @@ def apply_parametric_program(self, operations, **kwargs):
...
-            return self.qc.run(self.prog, memory_map=self._parameter_map)
+            return self.extract_samples(self.qc.run(self.prog))
 
         if self.circuit_hash is None:
             # Parametric compilation was set to False
             # Compile the program
             self._compiled_program = self.qc.compile(self.prog)
-            return self.qc.run(executable=self._compiled_program)
+            results = self.qc.run(executable=self._compiled_program)
+            return self.extract_samples(results)
```
Comment (AlbertMitjans) [inline pennylane_forest/qvm.py:251]: Careful because in this case the compiled program is not saved in the `self._compiled_program_dict` dictionary.

File: pennylane_forest/qvm.py [hunk 16]
```diff
@@ -244,8 +257,11 @@ def generate_samples(self):
 
         # The program has been compiled, store as the latest compiled program
         self._compiled_program = self._compiled_program_dict[self.circuit_hash]
-        samples = self.qc.run(executable=self._compiled_program, memory_map=self._parameter_map)
-        return samples
+        results = self.qc.run(executable=self._compiled_program)
+        return self.extract_samples(results)
+
+    def extract_samples(self, execution_results):
```
Comment (AlbertMitjans) [inline pennylane_forest/qvm.py:263]: Add docstring

**Review by AlbertMitjans — COMMENTED** [review]
Left some comments in the code. Still need to look at the tests.

File: pennylane_forest/qpu.py [hunk 17]
```diff
@@ -19,29 +19,24 @@
...
+from pyquil.paulis import sI, sZ
 from pyquil.quil import Program
 from pyquil.quilbase import Gate
 
 from pennylane.operation import Tensor
 
-from typing import List
-
-from ._version import __version__
 from .qvm import QVMDevice
```
Comment (AlbertMitjans) [inline pennylane_forest/qpu.py:40]: It is weird that the `QPUDevice` inherits from the `QVMDevice`... In the future we could either move the common methods to the `ForestDevice` class or create a new class

**Comment by AlbertMitjans** [comment]
The `pennylane_forest/device.py` file needs reformatting.

Please run `black -l 100 pennylane_forest/`.

File: README.rst (hunk 4 above)
Comment (MarquessV) [reply README.rst:60]: Should I just go ahead and remove this whole bullet point then?

File: doc/devices/qpu.rst (hunk 8 above)
Comment (MarquessV) [reply doc/devices/qpu.rst:86]: Good catch. These kwargs were removed but I didn't update the note to properly reflect that. I've corrected the note in both places.

File: doc/index.rst (hunk 13 above)
Comment (MarquessV) [reply doc/index.rst:52]: Looks like the `index-card` directive no longer supports figures. None of the [available directives in the xanadu-sphinx-theme](https://github.com/XanaduAI/xanadu-sphinx-theme#directives) are a perfect match for what I would imagine was being rendered before. There is a reference to a "Related Demo" directive, but it doesn't look like that's been finished yet. What would you recommend?

File: pennylane_forest/qpu.py (hunk 14 above)
Comment (MarquessV) [reply pennylane_forest/qpu.py:72]: Good catch.

File: pennylane_forest/qvm.py (hunk 15 above)
Comment (MarquessV) [reply pennylane_forest/qvm.py:251]: That is how it was before too. It looks like the original intent was to not cache the compiled programs if parametric compilation was disabled.

The [`circuit_hash` property returns `None` if parametric compilation is disabled](https://github.com/PennyLaneAI/pennylane-forest/pull/107/files/788b3cc2f73e0bf17700882328200802d89e644b#diff-e54edc4229030c98c5b70918872608e43667b52ebfafa4f6c3d93996eea94466R267).

and the [`_compiled_program_dict` is only initialized if parametric compilation is enabled](https://github.com/PennyLaneAI/pennylane-forest/pull/107/files/788b3cc2f73e0bf17700882328200802d89e644b#diff-e54edc4229030c98c5b70918872608e43667b52ebfafa4f6c3d93996eea94466R89).

Happy to update it if that's not what we want anymore, but want to make sure we're on the same page.

File: pennylane_forest/device.py [hunk 18]
```diff
@@ -176,18 +172,8 @@ def __init__(self, wires, shots=1000, **kwargs):
         self.reset()
```
[the remark this answers is not in the document]
Comment (MarquessV) [reply pennylane_forest/device.py:172]: done

File: pennylane_forest/wavefunction.py [hunk 19]
```diff
@@ -72,8 +61,7 @@ class WavefunctionDevice(ForestDevice):
 
     def __init__(self, wires, *, shots=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (MarquessV) [reply pennylane_forest/wavefunction.py]: done - removed it from the other child classes of `ForestDevice` as well.

File: pennylane_forest/qpu.py (hunk 17 above)
Comment (MarquessV) [reply pennylane_forest/qpu.py:40]: Agreed, that caught me off guard as well. I can create a followup issue if you'd like.

File: pennylane_forest/qpu.py [hunk 20]
```diff
@@ -246,17 +221,19 @@ def expval(self, observable):
                         label, p00=self.readout_error[0], p11=self.readout_error[1]
                     )
 
+            prep_prog.wrap_in_numshots_loop(self.shots)
```
Comment (MarquessV) [inline pennylane_forest/qpu.py:224]: @AlbertMitjans RE: https://github.com/PennyLaneAI/pennylane-forest/pull/107#discussion_r994449879

Where my confusion came from was that I wasn't sure how this code was working before. The `Program` wasn't being wrapped in a numshots loop and the number of shots wasn't being set on the Experiment either. Upon further investigation it turns out in `pyQuil` v2 an `Experiment` would default to 10k shots if it wasn't set on the `Program` or the `Experiment`.

Before, the code wasn't respecting the number of shots set by the user but it still passed all the tests because it defaulted to 10k shots. Since `Experiment`s no longer set shots in `pyQuil` v3 the program was only running a single time, causing the tests to fail due to inaccuracy and revealing the bug 🎊.

Given this, I removed the comment.

File: pennylane_forest/qvm.py (hunk 15 above)
Comment (AlbertMitjans) [reply pennylane_forest/qvm.py:252]: Oops my bad, you're right. Given that we are using the `circuit_hash` as dictionary keys, we can't cache it when it is None.

File: pennylane_forest/device.py [hunk 21]
```diff
@@ -36,22 +36,37 @@
...
 from pyquil import Program
-from pyquil.api._base_connection import ForestConnection
-from pyquil.api._config import PyquilConfig
-
 from pyquil.quil import DefGate
-from pyquil.gates import X, Y, Z, H, PHASE, RX, RY, RZ, CZ, SWAP, CNOT, S, T, CSWAP, I
-
-# following gates are not supported by PennyLane
-from pyquil.gates import CPHASE00, CPHASE01, CPHASE10, CPHASE, CCNOT, ISWAP, PSWAP
+from pyquil.gates import (
```
Comment (AlbertMitjans) [inline pennylane_forest/device.py:40]: Missing `CPHASE` import

File: doc/index.rst (hunk 13 above)
Comment (AlbertMitjans) [reply doc/index.rst:52]: I would just recommend removing the figure. But maybe @josh146 has a better solution.

File: README.rst (hunk 4 above)
Comment (AlbertMitjans) [reply README.rst:60]: Sure!

File: pennylane_forest/qpu.py [hunk 22]
```diff
@@ -66,28 +61,21 @@ class QPUDevice(QVMDevice):
...
+        execution_timeout (int): number of seconds to wait for a response from the QVM (default 10).
         parametric_compilation (bool): a boolean value of whether or not to use parametric
             compilation.
     """
     name = "Forest QPU Device"
     short_name = "forest.qpu"
     observables = {"PauliX", "PauliY", "PauliZ", "Identity", "Hadamard", "Hermitian"}
+    pennylane_requires = ">=0.15.0"
 
     def __init__(
```
Comment (AlbertMitjans) [inline pennylane_forest/qpu.py:80]: The `QPUDevice.__init__` contains duplicated code from the `QVMDevice.__init__`

Comment (AlbertMitjans) [reply pennylane_forest/qpu.py:80]: When calling super, use:
```python
super().__init__(wires, shots, **kwargs)
```

Instead of:
```python
super(QVMDevice, self).__init__(wires, shots, **kwargs)
```

**Comment by AlbertMitjans** [comment]
Remove `pytest.mark.xfail` from the `converter.py` tests.

**Comment by AlbertMitjans** [comment]
Do you know if I can run the tests locally on a M1 Mac? The docker image's platform is for linux and is not compatible with mac.

File: doc/index.rst (hunk 13 above)
Comment (MarquessV) [reply doc/index.rst:52]: I went ahead with that for now (and added a description). Happy to change it later if Josh has a suggestion!

File: pennylane_forest/device.py (hunk 21 above)
Comment (MarquessV) [reply pennylane_forest/device.py:40]: Added it in

File: pennylane_forest/qpu.py (hunk 22 above)
Comment (MarquessV) [reply pennylane_forest/qpu.py:80]: The way `super` is being used in `QPUDevice` is related to why the code is duplicated between `QPUDevice.__init__` and `QVMDevice.__init__`. 

For example, `QPUDevice` and `QVMDevice` initialize `self.qc` differently. If `QPUDevice` were to call `QVMDevice.__init__`, then it would effectively clobber the `qc` it had setup for itself. Because of this, it bypasses the `QVMDevice` by calling it's superclasses initializer (`ForestDevice.__init__`):

```python
super(QVMDevice, self).__init__(wires, shots, **kwargs)
```

Since they can't share an initializer, `QPUDevice` also has to duplicate all the stuff about initialization it shares in common with `QVMDevice`

As discussed in another thread, this is why `QPUDevice` shouldn't inherit from `QVMDevice`. The right move here is to refactor the shared logic into `ForestDevice` that both devices inherit from. I've gone ahead and started down that path since it's likely to cause more confusion until it's cleaned up.

File: tests/test_converter.py [hunk 23]
```diff
@@ -10,7 +10,6 @@
 from pennylane_forest.converter import *
 
 
-@pytest.mark.xfail
```
Comment (MarquessV) [inline tests/test_converter.py:13]: The majority of the converter tests pass, with the exception of 6, each of which fail with:

```
AttributeError: 'QNode' object has no attribute 'jacobian'
```

One example of this is on line 1233:

```
assert np.array_equal(qml.jacobian([angle]), circuit_reg.jacobian([angle]))
```

This seems like a change on the PennyLane side, any idea on what we need to do to fix?

File: pennylane_forest/qpu.py (hunk 22 above)
Comment (AlbertMitjans) [reply pennylane_forest/qpu.py:80]: Thanks for the explanation! I didn't realise we were calling `super(QVMDevice)` instead of `super(QPUDevice)`.

Comment (AlbertMitjans) [reply pennylane_forest/qpu.py:80]: Then we can also remove the `kwargs` from here 😄

File: tests/test_converter.py (hunk 23 above)
Comment (AlbertMitjans) [reply tests/test_converter.py:13]: Yes!

You need to use:
```python
assert np.array_equal(qml.jacobian(circuit)(angle), qml.jacobian(circuit_reg)(angle))
```

Instead of:
```python
assert np.array_equal(qml.jacobian([angle]), circuit_reg.jacobian([angle]))
```

Comment (MarquessV) [reply tests/test_converter.py:13]: That did the trick, thanks!

File: pennylane_forest/qc.py [hunk 24]
```diff
@@ -0,0 +1,407 @@
...
+        if self.parametric_compilation:
+            self._circuit_hash = circuit.graph.hash
+        return super().execute(circuit, **kwargs)
+
+    def generate_samples(self):
+        if self.parametric_compilation:
+            for region, value in self._parameter_map.items():
+                self.prog.write_memory(region_name=region, value=value)
+
+        if "pyqvm" in self.qc.name:
```
Comment (MarquessV) [inline pennylane_forest/qc.py:255]: This might not be necessary anymore, but I want to be able to run the pyqvm test suite before deleting it (along with moving the special case of `pyqvm`s in `apply` to the `QVMDevice`)

File: pennylane_forest/qpu.py [hunk 25]
```diff
@@ -190,7 +163,7 @@ def __init__(
         self.calibrate_readout = calibrate_readout
         self.wiring = {i: q for i, q in enumerate(self.qc.qubits())}
 
-    def expval(self, observable):
+    def expval(self, observable, shot_range=None, bin_size=None):
```
[the remark this answers is not in the document]
Comment (MarquessV) [reply pennylane_forest/qpu.py:118]: I took a brief look at doing this while refactoring the devices, but quickly ran into a wall when I realized that the logic in `expval` is dependent on the `observable` parameter to generate the right `PauliTerm` for the `Experiment` passed to `measure_observables`.

**Comment by AlbertMitjans** [comment]
I just realised that the github workflows run the tests using the `quilc` version 1.22.0. 

Probably that is why some of the tests are passing locally but failing here.

I think we should change the following line in `.github/workflows/tests.yml` from:
```
      - name: Run Forest Quilc
        run: docker run --rm -d -p 5555:5555 rigetti/quilc:1.22.0 -R
```
to
```
      - name: Run Forest Quilc
        run: docker run --rm -d -p 5555:5555 rigetti/quilc -S
```

**Comment by AlbertMitjans** [comment]
Update minimum pennylane version to `0.18` (ISWAP gate was added in the 0.18 release)

File: setup.py [hunk 26]
```diff
@@ -7,7 +7,7 @@
     version = f.readlines()[-1].split()[-1].strip("\"'")
 
 
-requirements = ["pyquil>=2.16,<2.28.3", "pennylane>=0.17"]
+requirements = ["pyquil>=3.3.1,<4.0.0", "qcs-api-client>=0.20.13<0.22.0", "pennylane>=0.17"]
```
Comment (AlbertMitjans) [inline setup.py:10]:
```suggestion
requirements = ["pyquil>=3.3.1,<4.0.0", "qcs-api-client>=0.20.13<0.22.0", "pennylane>=0.18"]

```

File: requirements.txt [hunk 27]
```diff
@@ -1,4 +1,5 @@
-pyquil>=2.16,<2.28.3
+pyquil>=3.0.0,<4.0.0
 pennylane>=0.17
```
Comment (AlbertMitjans) [inline requirements.txt:2]:
```suggestion
pennylane>=0.18
```

**Review by AlbertMitjans — COMMENTED** [review]
Update minimum pennylane version to `0.18` (ISWAP gate was added in the 0.18 release)

File: pennylane_forest/converter.py [hunk 28]
```diff
@@ -1,7 +1,5 @@
 import warnings
 from collections.abc import Sequence
-import pkg_resources
-from packaging import version
 
 import numpy as np
 import pennylane as qml
```
Comment (AlbertMitjans) [inline pennylane_forest/converter.py:5]: Update `pyquil_inv_operation_map` dictionary with:
```python
{
  "S": qml.S,
  "T": qml.T,
  "CCNOT": qml.Toffoli,
  "CSWAP": qml.CSWAP,
  "ISWAP": qml.ISWAP,
  "PSWAP": qml.PSWAP
}
```

And remove `S, T, CCNOT, CSWAP, ISWAP, PSWAP` gates from `pennylane_forest/ops.py`. Leave the `CPHASE` gate because I realised it has a bit more functionality than the `qml.ControlledPhase` gate in PennyLane.

File: pennylane_forest/converter.py [hunk 29]
```diff
@@ -1,7 +1,5 @@
 import warnings
```
Comment (AlbertMitjans) [inline pennylane_forest/converter.py:1]: In this file there is a line which uses the `pyquil.simulation` submodule:
```python
_matrix_dictionary = pyquil.simulation.matrices.QUANTUM_GATES
```

For some reason this now crashes the tests. I realised that in the current pyquil version, `simulation` is not imported top level, that is why Python can't find it.

This can be solved by importing simulation directly:
```python
import pyquil.simulation as simulation
```

File: pennylane_forest/converter.py (hunk 28 above)
Comment (AlbertMitjans) [reply pennylane_forest/converter.py:5]: The `plf.ops.CCNOT`, `plf.ops.CSWAP`... gates are used more than once in this file.

They all need to be replaced by `qml.Toffoli`, `qml.CSWAP`...

File: pennylane_forest/qc.py [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+            Setting this to ``True`` results in a significantly faster expectation value
+            evaluation when the number of shots is larger than ~1000.
+
+    Keyword args:
+        compiler_timeout (int): number of seconds to wait for a response from quilc (default 10).
+        execution_timeout (int): number of seconds to wait for a response from the QVM (default 10).
+        parametric_compilation (bool): a boolean value of whether or not to use parametric
+            compilation.
+    """
+    pennylane_requires = ">=0.17"
```
Comment (AlbertMitjans) [inline pennylane_forest/qc.py:64]:
```suggestion
    pennylane_requires = ">=0.18"
```

File: pennylane_forest/qc.py [hunk 31]
```diff
@@ -0,0 +1,296 @@
...
+            raise ValueError(
+                f"Device has a fixed number of {self.num_wires} qubits and "
+                f"cannot be created with {len(wires)} wires."
+            )
+
+        self.wiring = dict(enumerate(self.qc.qubits()))
+        self.active_reset = active_reset
+
+        super().__init__(wires, shots)
+        self.reset()
```
Comment (AlbertMitjans) [inline pennylane_forest/qc.py:121]: `self.reset()` is already called in `super().__init__`

File: pennylane_forest/qc.py [hunk 32]
```diff
@@ -0,0 +1,296 @@
...
+
+    @abstractmethod
+    def get_qc(self, device, **kwargs) -> QuantumComputer:
+        """Initializes and returns a pyQuil QuantumComputer that can run quantum programs"""
+
+    @property
+    def circuit_hash(self):
```
Comment (AlbertMitjans) [inline pennylane_forest/qc.py:128]: Add docstring

File: pennylane_forest/device.py [hunk 33]
```diff
@@ -191,17 +189,16 @@ def __init__(self, wires, shots=1000):
         super().__init__(wires, shots)
         self.reset()
```
Comment (AlbertMitjans) [inline pennylane_forest/device.py:190]: To avoid the CodeFactor warning:
```
Attribute 'prog' defined outside __init__ (attribute-defined-outside-init)
```

We could do:
```suggestion
        self.prog = Program()
        self._state = None
```

File: pennylane_forest/qc.py [hunk 34]
```diff
@@ -0,0 +1,296 @@
...
+                    # of the corresponding operation
+                    par.append(self._parameter_reference_map[parameter_string])
+                else:
+                    par.append(param)
+
+            prog += self._operation_map[operation.name](*par, *device_wires.labels)
+
+        return prog
+
+    def execute(self, circuit, **kwargs):
```
Comment (AlbertMitjans) [inline pennylane_forest/qc.py:245]: Add docstring

File: pennylane_forest/qc.py [hunk 35]
```diff
@@ -0,0 +1,296 @@
...
+        if self.parametric_compilation:
+            self._circuit_hash = circuit.graph.hash
+        return super().execute(circuit, **kwargs)
+
+    def generate_samples(self):
```
Comment (AlbertMitjans) [inline pennylane_forest/qc.py:250]: Add docstring

**Review by AlbertMitjans — COMMENTED** [review]
Amazing job with the refactor!

Added some coments.

File: pennylane_forest/qc.py (hunk 30 above)
Comment (MarquessV) [reply pennylane_forest/qc.py]: Set the min version of pennylane to `0.18` everywhere.

File: pennylane_forest/qc.py (hunk 31 above)
Comment (MarquessV) [reply pennylane_forest/qc.py]: Good catch, removed it.

File: pennylane_forest/qc.py (hunk 35 above)
Comment (MarquessV) [reply pennylane_forest/qc.py:257]: Added all the requested docstrings - definitely welcome to feedback on them.

File: pennylane_forest/device.py (hunk 33 above)
Comment (MarquessV) [reply pennylane_forest/device.py:190]: I think we have to have a valid `reset()` function since it is part of the `QubitDevice` API, so I don't think we can get around this rule. I added a comment to disable it.

Comment (AlbertMitjans) [reply pennylane_forest/device.py:190]: We can keep the `reset()` method, but I was suggesting using:
```python
self.prog = Program()
self._state = None
```
in the `__init__` function instead of calling `reset()`. This way the `prog` and `_state` attributes are defined in the `__init__` function and then CodeFactor won't complain anymore

File: pennylane_forest/device.py [hunk 36]
```diff
@@ -266,6 +266,7 @@ def apply_circuit_operations(self, operations):
 
     def reset(self):
         """Resets the device after the previous run."""
+        # pylint:disable=attribute-defined-outside-init
```
Comment (AlbertMitjans) [inline pennylane_forest/device.py:269]: As I mentioned earlier: if we use
```python
self.prog = Program()
self._state = None
```
instead of calling `reset()` in the `__init__` function, this pylint error won't happen.

**Comment by AlbertMitjans** [comment]
There was a timeout error in the `test_qvm.py:test_qubit_unitary` test. Should we increase the timeout threshold?

File: pennylane_forest/qpu.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane_forest/qpu.py:118]: I would leave this as it is for now. We can create a follow-up issue saying that this should be optimised.

File: pennylane_forest/device.py (hunk 36 above)
Comment (MarquessV) [reply pennylane_forest/device.py]: Okay, initialized those in `__init__`

File: pennylane_forest/qpu.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane_forest/qpu.py:118]: @antalszava

[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane_forest/qpu.py:118]: The QPU device allows operator estimation and uses `measure_observables` from PyQuil under the hood. This feature in the plugin is, however, not compatible with parametric compilation. Therefore, once parametric compilation is turned on, we're falling back to PennyLane's implementation using `generate_samples`.

A major shortcoming hinted at above too is that when doing operator estimation, we're effectively generating samples using PennyLane's pipeline **and** using `measure_observables` - the first of which should not be done.

See https://github.com/PennyLaneAI/pennylane-forest/issues/45 that has just been re-opened.

**Review by AlbertMitjans — APPROVED** [review]
Approved!
